libs/wire-api/src/Wire/API/OAuth.hs

```diff
@@ -0,0 +1,596 @@
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthScopes)
+
+instance ToSchema OAuthScopes where
+  schema = OAuthScopes <$> (oauthScopesToText . unOAuthScopes) .= withParser schema oauthScopeParser
+    where
+      oauthScopesToText :: Set OAuthScope -> Text
+      oauthScopesToText = T.intercalate " " . fmap (cs . toByteString') . Set.toList
+
+      oauthScopeParser :: Text -> A.Parser (Set OAuthScope)
+      oauthScopeParser "" = pure Set.empty
+      oauthScopeParser scope =
+        pure $ (not . T.null) `filter` T.splitOn " " scope & maybe Set.empty Set.fromList . mapM (fromByteString' . cs)
+
+data NewOAuthAuthCode = NewOAuthAuthCode
+  { noacClientId :: OAuthClientId,
+    noacScope :: OAuthScopes,
+    noacResponseType :: OAuthResponseType,
+    noacRedirectUri :: RedirectUrl,
+    noacState :: Text
+  }
+  deriving (Eq, Show, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema NewOAuthAuthCode)
+
+instance ToSchema NewOAuthAuthCode where
+  schema =
+    object "NewOAuthAuthCode" $
+      NewOAuthAuthCode
+        <$> noacClientId .= fieldWithDocModifier "client_id" clientIdDescription schema
+        <*> noacScope .= fieldWithDocModifier "scope" scopeDescription schema
+        <*> noacResponseType .= fieldWithDocModifier "response_type" responseTypeDescription schema
+        <*> noacRedirectUri .= fieldWithDocModifier "redirect_uri" redirectUriDescription schema
+        <*> noacState .= fieldWithDocModifier "state" stateDescription schema
+    where
+      clientIdDescription = description ?~ "The ID of the OAuth client"
+      scopeDescription = description ?~ "The scopes which are requested to get authorization for, separated by a space"
+      responseTypeDescription = description ?~ "Indicates which authorization flow to use. Use `code` for authorization code flow."
+      redirectUriDescription = description ?~ "The URL to which to redirect the browser after authorization has been granted by the user."
+      stateDescription = description ?~ "An opaque value used by the client to maintain state between the request and callback. The authorization server includes this value when redirecting the user-agent back to the client.  The parameter SHOULD be used for preventing cross-site request forgery"
+
+newtype OAuthAuthCode = OAuthAuthCode {unOAuthAuthCode :: AsciiBase16}
+  deriving (Eq, Generic)
+
+instance Show OAuthAuthCode where
+  show _ = "<OAuthAuthCode>"
+
+instance ToSchema OAuthAuthCode where
+  schema = (toText . unOAuthAuthCode) .= parsedText "OAuthAuthCode" (fmap OAuthAuthCode . validateBase16)
+
+instance ToByteString OAuthAuthCode where
+  builder = builder . unOAuthAuthCode
+
+instance FromByteString OAuthAuthCode where
+  parser = OAuthAuthCode <$> parser
+
+instance FromHttpApiData OAuthAuthCode where
+  parseQueryParam = bimap cs OAuthAuthCode . validateBase16 . cs
+
+instance ToHttpApiData OAuthAuthCode where
+  toQueryParam = toText . unOAuthAuthCode
```

Comment:
is this needed?  if not, it would be nice to have it removed so we don't accidentally use it to leak the code.

Reply:
We need this because the code is returned via a query parameter in the redirect url (in the location header)